doc/source/dev/data_types.md

```diff
@@ -404,8 +404,8 @@ class BasicComposite(Text):
 These files can be specified on the command line in the following fashion:
 
 ```xml
-<command>someTool.sh '$input1' '${os.path.join(input1.extra_files_path, 'results.txt')}'
-'${os.path.join(input1.extra_files_path, 'results.dat')}' '$output1'</command>
+<command>someTool.sh '$input1' '${os.path.join($input1.extra_files_path, 'results.txt')}'
```

Comment:
Are you positive you need `$input1` here?  I was also almost certain that once in the python context (the outer `${....}`, you wouldn't need this.

Reply:
Not 100% sure .. so I re-added the test that should fail (this time with out the typo .. I forgot the closing `}` the first time, ie in https://github.com/galaxyproject/galaxy/commit/f065a38b2b2ae434b2166e2f84f08ecb0883e080).

Reply:
So the framework tests are failing as expected. If ran in planemo the full Traceback is:

```
Traceback (most recent call last):
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/__init__.py", line 1053, in unicodify
    value = text_type(value)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596627208_607981_46507.py", line 89, in respond
AttributeError: 'builtin_function_or_method' object has no attribute 'extra_files_path'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/__init__.py", line 1055, in unicodify
    value = str(value)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596627208_607981_46507.py", line 89, in respond
AttributeError: 'builtin_function_or_method' object has no attribute 'extra_files_path'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/template.py", line 67, in fill_template
    return unicodify(t)
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/__init__.py", line 1061, in unicodify
    msg = "Value '%s' could not be coerced to Unicode: %s('%s')" % (value, type(e).__name__, e)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596627208_607981_46507.py", line 89, in respond
AttributeError: 'builtin_function_or_method' object has no attribute 'extra_files_path'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/jobs/runners/__init__.py", line 236, in prepare_job
    job_wrapper.prepare()
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/jobs/__init__.py", line 1107, in prepare
    self.command_line, self.extra_filenames, self.environment_variables = tool_evaluator.build()
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/tools/evaluation.py", line 475, in build
    raise e
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/tools/evaluation.py", line 471, in build
    self.__build_command_line()
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/tools/evaluation.py", line 496, in __build_command_line
    command_line = fill_template(command, context=param_dict, python_template_version=self.tool.python_template_version)
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/template.py", line 111, in fill_template
    python_template_version=python_template_version,
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/template.py", line 113, in fill_template
    raise first_exception or e
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/template.py", line 67, in fill_template
    return unicodify(t)
  File "/tmp/tmpkifhuuke/galaxy-dev/lib/galaxy/util/__init__.py", line 1061, in unicodify
    msg = "Value '%s' could not be coerced to Unicode: %s('%s')" % (value, type(e).__name__, e)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596627208_5605676_14480.py", line 87, in respond
AttributeError: 'builtin_function_or_method' object has no attribute 'extra_files_path'
FAIL
```

Reply:
Are you sure that the tool you tested used `input1`, not `input` (and then `input` might be the python builtin)? But okay, that should be reason enough to use the `$` variant.

Reply:
You are right, input seems to be a builtin. If I change the tool input name to something else (eg `blahfasel`) I get 

```
Traceback (most recent call last):
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/__init__.py", line 1053, in unicodify
    value = text_type(value)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596628047_2029321_19077.py", line 89, in respond
NameError: name 'blahfasel' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/__init__.py", line 1055, in unicodify
    value = str(value)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596628047_2029321_19077.py", line 89, in respond
NameError: name 'blahfasel' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/template.py", line 67, in fill_template
    return unicodify(t)
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/__init__.py", line 1061, in unicodify
    msg = "Value '%s' could not be coerced to Unicode: %s('%s')" % (value, type(e).__name__, e)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596628047_2029321_19077.py", line 89, in respond
NameError: name 'blahfasel' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/jobs/runners/__init__.py", line 236, in prepare_job
    job_wrapper.prepare()
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/jobs/__init__.py", line 1107, in prepare
    self.command_line, self.extra_filenames, self.environment_variables = tool_evaluator.build()
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/tools/evaluation.py", line 475, in build
    raise e
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/tools/evaluation.py", line 471, in build
    self.__build_command_line()
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/tools/evaluation.py", line 496, in __build_command_line
    command_line = fill_template(command, context=param_dict, python_template_version=self.tool.python_template_version)
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/template.py", line 111, in fill_template
    python_template_version=python_template_version,
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/template.py", line 113, in fill_template
    raise first_exception or e
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/template.py", line 67, in fill_template
    return unicodify(t)
  File "/tmp/tmpb5jlyh30/galaxy-dev/lib/galaxy/util/__init__.py", line 1061, in unicodify
    msg = "Value '%s' could not be coerced to Unicode: %s('%s')" % (value, type(e).__name__, e)
  File "/home/berntm/.planemo/gx_venv_3.7_release_20.05/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596628047_1631348_56014.py", line 87, in respond
NameError: name 'blahfasel' is not defined
FAIL
```

Reply:
With the extra `$` it works.

Reply:
Yeah, sure enough, just tried it locally, too.  This definitely did not work like I expected it to.